[HLSL] Implement `WaveReadLaneAt` intrinsic

inbelic opened the pull request:
- create a clang built-in in Builtins.td
    - add semantic checking in SemaHLSL.cpp
    - link the WaveReadLaneAt api in hlsl_intrinsics.h
    - add lowering to spirv backend op GroupNonUniformShuffle
      with Scope = 2 (Group) in SPIRVInstructionSelector.cpp
    - add WaveReadLaneAt intrinsic to IntrinsicsDirectX.td and mapping
      to DXIL.td

    - add tests for HLSL intrinsic lowering to spirv intrinsic in
      WaveReadLaneAt.hlsl
    - add tests for sema checks in WaveReadLaneAt-errors.hlsl
    - add spir-v backend tests in WaveReadLaneAt.ll
    - add test to show scalar dxil lowering functionality

    - note that this doesn't include support for the scalarizer to
      handle WaveReadLaneAt will be added in a future pr

This is the first part #70104

inbelic committed:
[HLSL] Implement `WaveReadLaneAt` intrinsic

    - create a clang built-in in Builtins.td
    - add semantic checking in SemaHLSL.cpp
    - link the WaveReadLaneAt api in hlsl_intrinsics.h
    - add lowering to spirv backend op GroupNonUniformShuffle
      with Scope = 2 (Group) in SPIRVInstructionSelector.cpp

    - add tests for HLSL intrinsic lowering to spirv intrinsic in
      WaveReadLaneAt.hlsl
    - add tests for sema checks in WaveReadLaneAt-errors.hlsl
    - add spir-v backend tests in WaveReadLaneAt.ll

inbelic committed:
[DXIL] add WaveReadLaneAt intrinsic for scalars

    - add WaveReadLaneAt intrinsic to IntrinsicsDirectX.td and mapping
      to DXIL.td

    - add test to show scalar functionality

    - note that this doesn't include support for the scalarizer to
      handle this function will be added in a future pr

inbelic commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2653,6 +2653,21 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
+        .addUse(GR.getSPIRVTypeID(ResType))
+        .addUse(I.getOperand(2).getReg())
+        .addUse(I.getOperand(3).getReg())
+        .addUse(GR.getOrCreateConstInt(2, I, IntTy, TII));
```
IIUC, an HLSL wave corresponds to the SPIR-V workgroup scope, which is denoted as 2.

Keenuts replied:
Looks like the correct value is 3? (cross-lane interaction, not cross-wave).
For the `OpGroupNonUniformElect`, the value MUST be 3 (spec)
DXC also generates both with a scope of 3.

inbelic replied:
Right, it should facilitate communication within the wave not between waves. Thanks.

Keenuts commented on clang/test/CodeGenHLSL/builtins/WaveReadLaneAt.hlsl:
```diff
@@ -0,0 +1,40 @@
+int test_int(int expr, uint idx) {
+  // CHECK-SPIRV: %[[#entry_tok:]] = call token @llvm.experimental.convergence.entry()
+
+  // CHECK-SPIRV:  %[[RET:.*]] = call [[TY:.*]] @llvm.spv.wave.read.lane.at.i32([[TY]] %[[#]], i32 %[[#]])
```
Not sure if that's done on other wave intrinsics, but we might want to check that the `convergencectrl` attribute is attached to it.

Keenuts reviewed (commented):
Thanks for this! Some minor changes, otherwise looks good 😊

inbelic committed:
review comments

    - add check for "convergencectrl" token in hlsl -> spirv intrinsic
    - correct the execution scope of the spirv instruction, add
      description
    - fix typo

s-perron commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,28 @@
+entry:
+  %0 = call token @llvm.experimental.convergence.entry()
+; CHECK:   %[[#ret:]] = OpGroupNonUniformShuffle %[[#f32]] %[[#expr]] %[[#idx]] %[[#scope]]
+  %1 = call float @llvm.spv.wave.read.lane.at(float %expr, i32 %idx) [ "convergencectrl"(token %0) ]
```
Could you add tests that check a few different expression/return types? It would be good to make sure that ti does not always return a float. It does not have to be as extensive as the DX test. Maybe just an 32-bit int and a vector of some sort.

inbelic committed:
review comments

    - extend spirv testcase to include i32 and vector of bools

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18835,6 +18835,22 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
+        Intrinsic::getName(CGM.getHLSLRuntime().getWaveReadLaneAtIntrinsic(),
+                           ArrayRef{OpExpr->getType()}, &CGM.getModule());
+    return EmitRuntimeCall(CGM.CreateRuntimeFunction(FT, name, {}, false, true),
+                           ArrayRef{OpExpr, OpIndex}, "hlsl.wave.read.lane.at");
```
The name we used for `BI__builtin_hlsl_wave_get_lane_index` was  `__hlsl_wave_get_lane_index`.  Why would we use periods here?

inbelic replied:
I think `__hlsl_wave_get_lane_index` is the odd one out. The other intrinsics follow the pattern of `hlsl.name`. Having changed to using one word `waveReadLaneAt` I think we can keep it consistent naming with `hlsl.waveReadLaneAt`.

I can change the name to `hlsl.waveGetLaneIndex` in the clean-up pr.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18835,6 +18835,22 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
+    std::string name =
+        Intrinsic::getName(CGM.getHLSLRuntime().getWaveReadLaneAtIntrinsic(),
+                           ArrayRef{OpExpr->getType()}, &CGM.getModule());
+    return EmitRuntimeCall(CGM.CreateRuntimeFunction(FT, name, {}, false, true),
```
put  the arg name for false and true via a comment  something like `/*BlahType=*/false, /*BlahType2=*/true)` but instead of Blah use the arg name in `CreateRuntimeFunction`

farzonl commented on llvm/include/llvm/IR/IntrinsicsDirectX.td:
```diff
@@ -83,6 +83,7 @@ def int_dx_umad : DefaultAttrsIntrinsic<[llvm_anyint_ty], [LLVMMatchType<0>, LLV
 def int_dx_normalize : DefaultAttrsIntrinsic<[LLVMMatchType<0>], [llvm_anyfloat_ty], [IntrNoMem]>;
 def int_dx_rsqrt  : DefaultAttrsIntrinsic<[llvm_anyfloat_ty], [LLVMMatchType<0>], [IntrNoMem]>;
 def int_dx_wave_is_first_lane : DefaultAttrsIntrinsic<[llvm_i1_ty], [], [IntrConvergent]>;
+def int_dx_wave_read_lane_at : DefaultAttrsIntrinsic<[llvm_any_ty], [LLVMMatchType<0>, llvm_i32_ty], [IntrConvergent]>;
```
I see `int_dx_wave_is_first_lane` doesn't have `IntrNoMem` either but I know `IntrNoMem` was required to scalarize other intrinsics and this is an intrinsic we want to scalarize. Is the `IntrConvergent` attribute sufficient?

inbelic replied:
My understanding was that `IntrNoMem` denotes there are no side-effects in the operation and that cross-lane communication would be a side-effect.

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -2653,6 +2653,22 @@ bool SPIRVInstructionSelector::selectIntrinsic(Register ResVReg,
         .addUse(GR.getOrCreateConstInt(3, I, IntTy, TII));
   }
+  case Intrinsic::spv_wave_read_lane_at: {
+    assert(I.getNumOperands() == 4);
```
Put this block in a select function. It makes it easier to see all the various case statements when scrolling through the intrinsic selector.

farzonl commented on llvm/test/CodeGen/DirectX/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,56 @@
+declare i16 @llvm.dx.wave.read.lane.at.i16(i16, i32) #1
+declare i32 @llvm.dx.wave.read.lane.at.i32(i32, i32) #1
+
+attributes #0 = { norecurse "hlsl.numthreads"="1,1,1" "hlsl.shader"="compute" "no-trapping-math"="true" "stack-protector-buffer-size"="8" }
```
delete the attributes. line 55 and 56

farzonl commented on llvm/test/CodeGen/DirectX/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,56 @@
+  ret i32 %ret
+}
+
+declare half @llvm.dx.wave.read.lane.at.f16(half, i32) #1
```
remove all the `#1`

farzonl commented on llvm/test/CodeGen/DirectX/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,56 @@
+define noundef i16 @wave_rla_i16(i16 noundef %expr, i32 noundef %idx) #0 {
+entry:
+; CHECK: call i16 @dx.op.waveReadLaneAt.i16(i32 117, i16 %expr, i32 %idx)
+  %ret = call i16 @llvm.dx.wave.read.lane.at.i16(i16 %expr, i32 %idx)
```
Now that i'm seeing how the intrinsic looks I think I would prefer it to be one word like it is on the dxop.  Dots are used to indicate namespaces and type information and doing it like this is making the intrinsic look confusing.
```suggestion
  %ret = call i16 @llvm.dx.waveReadLaneAt.i16(i16 %expr, i32 %idx)
```

farzonl commented on llvm/test/CodeGen/DirectX/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,56 @@
+  ret i16 %ret
+}
+
+define noundef i32 @wave_rla_i32(i32 noundef %expr, i32 noundef %idx) #0 {
```
remove all the `#0`

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,28 @@
+
+declare i32 @__hlsl_wave_get_lane_index() #1
+
+attributes #0 = { convergent norecurse "hlsl.numthreads"="1,1,1" "hlsl.shader"="compute" "no-trapping-math"="true" "stack-protector-buffer-size"="8" }
```
delete lines 22 to 28.

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,28 @@
+  ret void
+}
+
+declare i32 @__hlsl_wave_get_lane_index() #1
```
remove the  `#1`

farzonl replied:
I don't see this declare get used. I think it was a copy from a past test? You need to update the declare for `WaveReadLaneAt`.

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,28 @@
+; CHECK-DAG:   %[[#expr:]] = OpFunctionParameter %[[#f32]]
+; CHECK-DAG:   %[[#idx:]] = OpFunctionParameter %[[#uint]]
+
+define spir_func void @test_1(float %expr, i32 %idx) #0 {
```
remove the '#0`

farzonl commented on llvm/include/llvm/IR/IntrinsicsSPIRV.td:
```diff
@@ -82,5 +82,6 @@ let TargetPrefix = "spv" in {
     [llvm_anyint_ty, LLVMScalarOrSameVectorWidth<0, LLVMVectorElementType<0>>],
     [IntrNoMem, Commutative] >;
   def int_spv_wave_is_first_lane : DefaultAttrsIntrinsic<[llvm_i1_ty], [], [IntrConvergent]>;
+  def int_spv_wave_read_lane_at : DefaultAttrsIntrinsic<[llvm_any_ty], [LLVMMatchType<0>, llvm_i32_ty], [IntrConvergent]>;
```
`wave_read_lane_at` would be better as one word.

Probably should update `wave_is_first_lane` in a follow cleanup on pr.

inbelic replied:
Added a commit to change the name to `waveReadLaneAt` as described in below comment.

inbelic committed:
review comments

    - remove unneeded attributes from test functions
    - provide context to parameters in CGBuiltin
    - update intrinsic naming for clarity

inbelic committed:
review comments:

    - change spv/dx intrinsics naming from wave_read_lane_at to
      waveReadLaneAt

inbelic committed:
review comments:

    - add `IntrNoMem` to intrinsics to denote no memory access

llvm-beanz commented on clang/lib/Headers/hlsl/hlsl_intrinsics.h:
```diff
@@ -2015,6 +2015,13 @@ _HLSL_AVAILABILITY(shadermodel, 6.0)
+template <typename T>
+_HLSL_AVAILABILITY(shadermodel, 6.0)
+_HLSL_BUILTIN_ALIAS(__builtin_hlsl_wave_read_lane_at)
+    __attribute__((convergent)) T WaveReadLaneAt(T, int32_t);
```
Despite the documentation, this doesn't seem to work in DXC for arbitrary types:

https://godbolt.org/z/48brf1bre

@tex3d, gen_intrin_main says this intrinsic is valid on `any<>`, which seems to be any vector or scalar. Is that correct?

inbelic replied:
Fwiw, this is also true for the `select` intrinsic and we use the arbitrary type template.

tex3d replied:
`any<>` means scalar, vector, or matrix, with any component type valid with those (bool and basic numeric types).

llvm-beanz replied:
> Fwiw, this is also true for the `select` intrinsic and we use the arbitrary type template.

`select` is a little different, there are two overloads of it, one for numeric types and one for samplers. Really it should have worked with any type, but gen_intrin_main isn't expressive enough.

farzonl commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18835,6 +18835,24 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
+    std::string name =
+        Intrinsic::getName(CGM.getHLSLRuntime().getWaveReadLaneAtIntrinsic(),
+                           ArrayRef{OpExpr->getType()}, &CGM.getModule());
+    return EmitRuntimeCall(CGM.CreateRuntimeFunction(FT, name, {},
```
It will  we need to modify https://github.com/llvm/llvm-project/blob/848cb21ddca5dc4d8c52e487264936be8f1bfe8d/llvm/lib/Transforms/Scalar/Scalarizer.cpp#L748

to become:
```c++
if (isVectorIntrinsicWithScalarOpAtArg(ID, J) || TTI->isTargetIntrinsicWithScalarOpAtArg(ID, J)) 
```
by adding a new `TargetTransformInfo` api

inbelic committed:
review comments:

    - remove use of arbitrary type template to match the
      `gen_intrin_main` definition of any scalar or vector with a valid
      component type of bool and numeric types

    - correct type check to not allow a vector of ints

inbelic committed:
self review:

    - add support for selecting a non-side effect convergent intrinsic
    - update test case to not be optimized away

farzonl commented on llvm/lib/Target/SPIRV/SPIRVInstructionSelector.cpp:
```diff
@@ -428,6 +431,7 @@ bool SPIRVInstructionSelector::spvSelect(Register ResVReg,
 
   case TargetOpcode::G_INTRINSIC:
   case TargetOpcode::G_INTRINSIC_W_SIDE_EFFECTS:
+  case TargetOpcode::G_INTRINSIC_CONVERGENT:
```
Is there anything that distinguishes a `G_INTRINSIC_CONVERGENT` vs a `G_INTRINSIC`?  Seems like we are handling them the same way so wondering what happens with that extra information about convergence?

inbelic replied:
Afaict, not within this section of the codebase, it is used elsewhere though eg `GlobalISel`. But I think @Keenuts would be able to better elaborate.

s-perron replied:
The extra convergence information is being used when we generate the merge instructions. They should not change how the instrinsic itself is generated. The convergence token help us distinguish between:

```
int i;
for(i = 0; i < N; ++i) {
  if (cond(i)) {
    // The "active" lanes are those where cond was true in the same iteration.
    WaveReadLaneAt(i);
    break;
  }
}


for(i = 0; i < N; ++i) {
  if (cond(i))
    break;
}
// All lanes are active because they all reconverged at the end of the loop.
WaveReadLaneAt(i);
```

farzonl reviewed (approved):
LGTM

inbelic commented:
I will go ahead and merge this tomorrow if there are no other comments.

llvm-beanz commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -18835,6 +18835,24 @@ case Builtin::BI__builtin_hlsl_elementwise_isinf: {
+        false);
+
+    // Get overloaded name
+    std::string name =
+        Intrinsic::getName(CGM.getHLSLRuntime().getWaveReadLaneAtIntrinsic(),
+                           ArrayRef{OpExpr->getType()}, &CGM.getModule());
+    return EmitRuntimeCall(CGM.CreateRuntimeFunction(FT, name, {},
```
nit: https://llvm.org/docs/CodingStandards.html#name-types-functions-variables-and-enumerators-properly
```suggestion
    std::string Name =
        Intrinsic::getName(CGM.getHLSLRuntime().getWaveReadLaneAtIntrinsic(),
                           ArrayRef{OpExpr->getType()}, &CGM.getModule());
    return EmitRuntimeCall(CGM.CreateRuntimeFunction(FT, Name, {},
```

llvm-beanz commented on clang/lib/CodeGen/CGHLSLRuntime.h:
```diff
@@ -87,6 +87,7 @@ class CGHLSLRuntime {
   GENERATE_HLSL_INTRINSIC_FUNCTION(SDot, sdot)
   GENERATE_HLSL_INTRINSIC_FUNCTION(UDot, udot)
   GENERATE_HLSL_INTRINSIC_FUNCTION(WaveIsFirstLane, wave_is_first_lane)
+  GENERATE_HLSL_INTRINSIC_FUNCTION(WaveReadLaneAt, waveReadLaneAt)
```
nit: Intrinsic declarations are all `_` separated, not `camelCase`.

inbelic replied:
If we use `_` separated then the generated names will use `.` to separate them. So eg `wave_read_lane_at` becomes `llvm.dx.wave.read.lane.at`. Which we decided was quite confusing due to the confusion with a namespace. I will clean up the other intrinsic to cohere to this in a follow up pr.

llvm-beanz replied:
This isn't just a convention for our backend, this is a convention common across LLVM backends. For example look at the AMD GCN intrinsics like `llvm.amdgcn.raw.ptr.buffer.store`.

I can see an argument that the `_`s were placed in the wrong places, but I think this was the wrong decision. A better name might be `wave_readlaneat`, but `camelCase` isn't really used in intrinsic names.

@bogner & @farzonl, we should discuss this before we break with common LLVM convention.

llvm-beanz replied:
The equivalent AMD GPU intrinsic is named `int_amdgcn_readlane`, which translates to `llvm.amdgcn.readlane`. Having `dx.readlane` or `dx.wave.readlane` seems like a preferable naming approach.

farzonl replied:
This is my bad. Finn was using the llvm convention. I didn’t like the way it looked.

inbelic replied:
Updated to `wave_readlaneat`.

llvm-beanz commented on clang/test/SemaHLSL/BuiltIns/WaveReadLaneAt-errors.hlsl:
```diff
@@ -0,0 +1,26 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -verify-ignore-unexpected
```
I'm always suspicious when I see `verify-ignore-unexpected` especially if it doesn't specify a specific type. What diagnostics is this ignoring?

inbelic replied:
Good catch. There are no ignored diagnostics and it is not needed.

inbelic committed:
review comments:

    - fix variable name to be uppercase
    - remove unneeded flags from SemaHLSL testcase

inbelic committed:
review comments:

    - switch to wave_readlaneat to follow llvm naming conventions

llvm-beanz commented on clang/lib/CodeGen/CGHLSLRuntime.h:
```diff
@@ -87,6 +87,7 @@ class CGHLSLRuntime {
   GENERATE_HLSL_INTRINSIC_FUNCTION(SDot, sdot)
   GENERATE_HLSL_INTRINSIC_FUNCTION(UDot, udot)
   GENERATE_HLSL_INTRINSIC_FUNCTION(WaveIsFirstLane, wave_is_first_lane)
+  GENERATE_HLSL_INTRINSIC_FUNCTION(WaveReadLaneAt, wave_readlaneat)
```
It is probably worth naming this `wave_readlane` to align a bit with `amdgcn.readlane`. I'm not sure what the equivalent NV PTX intrinsic is... probably some crazy bundle of shuffles which confuses me.

llvm-beanz reviewed (commented):
One small nit, but otherwise looks good.

llvm-beanz commented on clang/lib/Sema/SemaHLSL.cpp:
```diff
@@ -1956,6 +1956,26 @@ bool SemaHLSL::CheckBuiltinFunctionCall(unsigned BuiltinID, CallExpr *TheCall) {
+    }
+
+    // Ensure return type is the same as the input expr type
+    ExprResult Expr = TheCall->getArg(0);
```
We should verify that the type of argument 0 is either a scalar or vector.

inbelic committed:
review comments:

    - change intrinsic names to `wave_readlane` to align with AMD
      implementations
    - update semantic check to ensure that only a scalar/vector is
      allowed
    - add test case to illustrate this

inbelic committed:
Merge branch 'main' into inbelic/wave-read-at-spirv

inbelic committed:
clang formatting

bogner reviewed (approved):
Looks good once Chris's comments and the clang-format issue are resolved.

inbelic committed:
self review:

    - the execution scope must be the second operand and not the final
      operand of the SPIRV instruction. this was missed as I did not
      realize that SPIRV-TOOLS was not enabled on my local machine but
      luckily got caught by the spir-v github tests

    - missing testcase for int64 in directx lowering and caught that
      Int64Ty was missing from the overload types of dxilop

bogner commented on clang/test/CodeGenHLSL/builtins/WaveReadLaneAt.hlsl:
It would probably be good to have some test coverage for int16, half, and double here

inbelic committed:
review comments:

    - add tests for int16, half and double to hlsl codegen
    - add test for vector of floats in spirv codegen to ensure vector
      register allocation

farzonl commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,44 @@
+define <4 x i1> @test_vbool(<4 x i1> %vbexpr, i32 %idx) {
+entry:
+; CHECK:   %[[#vbret:]] = OpGroupNonUniformShuffle %[[#v4_bool]] %[[#scope]] %[[#vbexpr]] %[[#idx3]]
+  %0 = call <4 x i1> @llvm.spv.wave.readlane.v4i1(<4 x i1> %vbexpr, i32 %idx)
```
if it isn't too much trouble could you do a `float4` test. Given our meeting today want to make sure vector float registers get allocated.

s-perron commented on llvm/test/CodeGen/SPIRV/hlsl-intrinsics/WaveReadLaneAt.ll:
```diff
@@ -0,0 +1,56 @@
+; RUN: llc -verify-machineinstrs -O0 -mtriple=spirv32v1.3-vulkan-unknown %s -o - | FileCheck %s
+; RUN: %if spirv-tools %{ llc -O0 -mtriple=spirv32v1.3-vulkan-unknown %s -o - -filetype=obj | spirv-val %}
```
Sorry I missed this in my review. The is the wrong triple to use. You should be using `spirv` not `spirv32`.

s-perron replied:
I don't think anything else needs to be updated. I'll open a PR to fix this myself.

s-perron replied:
https://github.com/llvm/llvm-project/pull/122738
